src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPullRequestHookProcessor.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    @Override
+    public void process(HookEventType hookEvent, String payload, BitbucketType instanceType, String origin, String serverUrl) {
+
+        final NativeServerPullRequestEvent pullRequestEvent;
+        try {
+            pullRequestEvent = JsonParser.toJava(payload, NativeServerPullRequestEvent.class);
+        } catch (final IOException e) {
+            LOGGER.log(Level.SEVERE, "Can not read hook payload", e);
+            return;
+        }
+
+        final SCMEvent.Type eventType;
+        switch (hookEvent) {
+            case SERVER_PULL_REQUEST_OPENED:
+                eventType = SCMEvent.Type.CREATED;
+                break;
+            case SERVER_PULL_REQUEST_DECLINED:
+            case SERVER_PULL_REQUEST_MERGED:
+                eventType = SCMEvent.Type.REMOVED;
+                break;
+            default:
+                LOGGER.log(Level.INFO, "Unknown hook event {0} received from Bitbucket Server", hookEvent);
+                return;
+        }
+
+        SCMHeadEvent.fireNow(new HeadEvent(serverUrl, eventType, pullRequestEvent, origin));
+    }
+
+    private static final class HeadEvent extends NativeServerHeadEvent<NativeServerPullRequestEvent> {
+        private HeadEvent(String serverUrl, Type type, NativeServerPullRequestEvent payload, String origin) {
+            super(serverUrl, type, payload, origin);
+        }
+
+        @Override
+        protected BitbucketServerRepository getRepository() {
+            return getPayload().getPullRequest().getDestination().getRepository();
+        }
+
+        @NonNull
+        @Override
+        protected Map<SCMHead, SCMRevision> heads(@NonNull BitbucketSCMSource source) {
+            if (!eventMatchesRepo(source)) {
+                return Collections.emptyMap();
+            }
+
+            final BitbucketSCMSourceContext ctx = contextOf(source);
+            if (!ctx.wantPRs()) {
+                return Collections.emptyMap(); // doesn't want PRs, nothing to do here
+            }
+
+            final BitbucketPullRequest pullRequest = getPayload().getPullRequest();
+            final BitbucketRepository sourceRepo = pullRequest.getSource().getRepository();
+            final SCMHeadOrigin headOrigin = source.originOf(sourceRepo.getOwnerName(), sourceRepo.getRepositoryName());
+            final Set<ChangeRequestCheckoutStrategy> strategies = headOrigin == SCMHeadOrigin.DEFAULT
+                ? ctx.originPRStrategies()
+                : ctx.forkPRStrategies();
+            final Map<SCMHead, SCMRevision> result = new HashMap<>(strategies.size());
+            for (final ChangeRequestCheckoutStrategy strategy : strategies) {
+                final String branchName = String.format("PR-%s%s", pullRequest.getId(),
+                    strategies.size() > 1 ? "-" + Ascii.toLowerCase(strategy.name()) : "");
+                final PullRequestSCMHead head = new PullRequestSCMHead(branchName, source.getRepoOwner(), source.getRepository(),
```

Comment:
you should use `pullRequest.getSource().getBranch().getName()` instead of `branchName` otherwise it will not trigger the PR build (see how PullRequestSCMHead is created in `com.cloudbees.jenkins.plugins.bitbucket.BitbucketSCMSource#retrievePullRequests`)

It will not trigger because it will fail the observerIncludes check in `jenkins.scm.api.trait.SCMSourceRequest#isExcluded` and the untrusted check in `jenkins.scm.api.SCMHeadEvent.Validated#observe`

I had this problem trying to have PR triggered builds when PR are from a branch in the same origin. 

You can check also that all the remaining calls to PullRequestSCMHead constructor uses pullRequest.getSource().getBranch().getName() instead of branchName 

The change that solve my issue look like:
```
final PullRequestSCMHead head = new PullRequestSCMHead(branchName, source.getRepoOwner(), source.getRepository(),
                    BitbucketRepositoryType.GIT, pullRequest.getSource().getBranch().getName(), pullRequest, headOrigin, strategy);
```

Reply:
Interesting. This seems to be the exact setup that's working for me. I don't think that we can "just" take the branch name, since this would almost certainly break PR builds, especially with respect to the checkout strategies.

Could you somehow provide more info about your setup? Maybe I can reproduce your problem.

/cc @casz

Reply:
Okay, I got the time to setup a Docker based test setup, using BB 5.13.1 and Jenkins 2.145. Did as you said and I wasn't able to reproduce the problem. Branches pop up almost instantaneously and are being built without issues.

Are you sure that the BB webhooks reached your Jenkins instance? What does the BitBucket webhook event log say (for my repo it is located at `Repository Settings -> Webhooks -> View Details on Jenkins Webhook in the Webhook list`: `http://<BB_URL>/plugins/servlet/webhooks/projects/FOO/repos/bar/details/1`)? What does the Jenkins Multibranch Pipeline Events Log say (for my test repo it is located at `http://<JENKINS_URL>/job/foo/job/bar/indexing/events`)?

Reply:
Hi, I'm sure that webhooks  reach my Jenkins without problems. tomorrow morning I can send you the details that you are asking. 
The problem is not with branches (those pop up without problems). The issue is with pull requests, those will only pop up after trigger a scan.

Reply:
Here are some logs and screenshots:

![screenshot_2018-10-22 webhooks - bitbucket](https://user-images.githubusercontent.com/1215104/47319147-40567080-d64e-11e8-9001-75cf38d5d54c.png)
![screenshot_2018-10-22 webhooks - bitbucket 2](https://user-images.githubusercontent.com/1215104/47319148-40567080-d64e-11e8-8af2-6b9b9d6c5e2c.png)
![screenshot_2018-10-22 webhooks - bitbucket 1](https://user-images.githubusercontent.com/1215104/47319149-40567080-d64e-11e8-8149-65f8a7c3e3b7.png)

Bitbucket Event Payload:

```json
{
  "eventKey": "repo:refs_changed",
  "date": "2018-10-22T20:41:28+0000",
  "actor": {
    "name": "tom",
    "emailAddress": "tom@bibbu.net",
    "id": 1,
    "displayName": "tom",
    "active": true,
    "slug": "tom",
    "type": "NORMAL"
  },
  "repository": {
    "slug": "bar",
    "id": 1,
    "name": "bar",
    "scmId": "git",
    "state": "AVAILABLE",
    "statusMessage": "Available",
    "forkable": true,
    "project": {
      "key": "FOO",
      "id": 1,
      "name": "Foo",
      "public": false,
      "type": "NORMAL"
    },
    "public": false
  },
  "changes": [
    {
      "ref": {
        "id": "refs/heads/feature/f4",
        "displayId": "feature/f4",
        "type": "BRANCH"
      },
      "refId": "refs/heads/feature/f4",
      "fromHash": "0000000000000000000000000000000000000000",
      "toHash": "fbccf6a08e1a5428740173511ec3592948254356",
      "type": "ADD"
    }
  ]
}
```

Multibranch Pipeline Events:
```
[Mon Oct 22 20:39:51 UTC 2018] Received com.cloudbees.jenkins.plugins.bitbucket.hooks.NativeServerPushHookProcessor$HeadEvent CREATED event from 172.19.0.2 ⇒ http://172.19.0.3:8080/bitbucket-scmsource-hook/notify with timestamp Mon Oct 22 20:39:51 UTC 2018
Connecting to http://bitbucket-5131:7990 using tom/******
Repository type: Git
Looking up FOO/bar for branches
Checking branch master from FOO/bar
Checking branch dev from FOO/bar
      ‘Jenkinsfile’ found
    Met criteria
Scheduled build for branch: dev

  2 branches were processed (query completed)
[Mon Oct 22 20:39:51 UTC 2018] com.cloudbees.jenkins.plugins.bitbucket.hooks.NativeServerPushHookProcessor$HeadEvent CREATED event from 172.19.0.2 ⇒ http://172.19.0.3:8080/bitbucket-scmsource-hook/notify with timestamp Mon Oct 22 20:39:51 UTC 2018 processed in 0.57 sec
[Mon Oct 22 20:41:28 UTC 2018] Received com.cloudbees.jenkins.plugins.bitbucket.hooks.NativeServerPushHookProcessor$HeadEvent CREATED event from 172.19.0.2 ⇒ http://172.19.0.3:8080/bitbucket-scmsource-hook/notify with timestamp Mon Oct 22 20:41:28 UTC 2018
Connecting to http://bitbucket-5131:7990 using tom/******
Repository type: Git
Looking up FOO/bar for branches
Checking branch master from FOO/bar
Checking branch feature/f4 from FOO/bar
      ‘Jenkinsfile’ found
    Met criteria
Scheduled build for branch: feature/f4

  2 branches were processed (query completed)
[Mon Oct 22 20:41:28 UTC 2018] com.cloudbees.jenkins.plugins.bitbucket.hooks.NativeServerPushHookProcessor$HeadEvent CREATED event from 172.19.0.2 ⇒ http://172.19.0.3:8080/bitbucket-scmsource-hook/notify with timestamp Mon Oct 22 20:41:28 UTC 2018 processed in 0.33 sec
```

Reply:
Doh! Forgot to open the PR. Okay, this doesn't work for PRs. You're right. So something has changed between Bitbuvket versions 5.9 (for which PRs work) and 5.13 (for which they do not) :unamused:

Reply:
Figured out why PRs don't work any longer. This is a regression that's been introduced by #114. So this has nothing to do with BB or Jenkins versions. That's why it's working great on our Jenkins instances, but not with the current version of this branch. We still use vanilla 2.2.12 plus this patch, which doesn't yet include #114.

Any thoughts on how to solve this? Will just using branchName here be okay and work for multiple merge strategies?

Reply:
The reason for #114 is because it was a initially was mistake from the author who last changed those lines.

So I believe you shouldn't change #114 and instead attempt to find out where the issue lies.